docs/RespLogging.md

```diff
@@ -0,0 +1,150 @@
+{
+    // this is deliberately not using SCARD
+    // (to put load on the inbound)
+    count += (await db.SetMembersAsync(testKey)).Length;
+}
+Console.WriteLine("all done");
+```
+
+## Logging RESP streams
+
+When this fails, it will not be obvious exactly who is to blame. However, we can ask for the data streams
+to be logged to the local file-system.
+
+**Obviously, this may leave data on disk, so this may present security concerns if used with production data; use
+this feature sparingly, and clean up after yourself!**
+
+``` c#
+// connect
+Console.WriteLine("Connecting...");
+var options = ConfigurationOptions.Parse(ConnectionString);
+LoggingTunnel.LogToDirectory(options, @"C:\Code\RedisLog"); // <=== added!
+await using var muxer = await ConnectionMultiplexer.ConnectAsync(options);
+...
+```
+
+This API is marked `[Obsolete]` simply to discourage usage, but you can ignore this warning once you
+understand what it is saying (using `#pragma warning disable CS0618` if necessary).
+
+This will update the `ConfigurationOptions` with a custom `Tunnel` that performs file-based mirroring
+of the RESP streams. If `Ssl` is enabled on the `ConfigurationOptions`, the `Tunnel` will *take over that responsibility*
+(so that the unencrypted data can be logged), and will *disable* `Ssl` on the `ConfigurationOptions` - but TLS
+will still be used correctly.
+
+If we run our code, we will see that 2 files are written per connection ("in" and "out"); if you are using RESP2 (the default),
+then 2 connections are usually established (one for regular "interactive" commands, and one for pub/sub messages), so this will
+typically create 4 files.
+
+## Validating RESP streams
+
+RESP is *mostly* text, so a quick eyeball can be achieved using any text tool; an "out" file will typically start:
+
+``` txt
+$6
+CLIENT
+$7
+SETNAME
+...
+```
+
+and an "in" file will typically start:
+
+``` txt
++OK
++OK
++OK
+...
+```
+
+This is the start of the handshakes for identifying the client to the redis server, and the server acknowledging this (if
+you have authentication enabled, there will be a `AUTH` command first, or `HELLO` on RESP3).
+
+If there is a failure, you obviously don't want to manually check these files. Instead, an API exists to validate RESP streams:
+
+``` c#
+var messages = await LoggingTunnel.ValidateAsync(@"C:\Code\RedisLog");
+Console.WriteLine($"{messages} RESP fragments validated");
+```
+
+If the RESP streams are *not* valid, an exception will provide further details.
+
+**An exception here is strong evidence that there is a fault either in the redis server, or an intermediate proxy**.
+
+Conversely, if the library reported a protocol failure but the validation step here *does not* report an error, then
+that is strong evidence of a library error; [**please report this**](https://github.com/StackExchange/StackExchange.Redis/issues/new) (with details).
+
+You can also *replay* the conversation locally, seeing the individual requests and responses:
```

Comment:
Could we "replay" files into the library itself for some tests (as opposed to running an actual Redis server locally)?

Reply:
there's a few problems there; we'd need to isolate away anything non-deterministic, which means:

1. anything involving multiple threads, including anything that is a side-effect of heartbeat and (due to timing) anything pub/sub on RESP2
2. anything random, such as the unique muxer identifier (we'd have to make it deterministic in for this scenario, or make it forgive the value received)
3. anything involving multiple connections (cluster)

If we had a concrete scenario where all of the above would be met, and replay would be useful: fine, but I don't think we should do anything preemptively.